src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,230 @@
+(hol : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (↑f : ℍ' → ℂ))
+(transf : f ∈ weakly_modular_form k Γ)
+(infinity : ∀ (A : SL(2, ℤ)), is_zero_at_im_infty (f ∣[k, A]))
+
+lemma is_modular_form_of_is_cusp_form {f : ℍ → ℂ}
+  (h : is_cusp_form k Γ f) : is_modular_form k Γ f :=
+{ hol := h.hol,
+  transf := h.transf,
+  infinity := λ (A : SL(2, ℤ)), filter.zero_at_filter_is_bounded_at_filter (h.infinity A)}
+
+/-- This is the space of cuspforms of weigth `k` and level `Γ` -/
+def space_of_cusp_forms_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2, ℤ)) :
+  submodule ℂ (ℍ → ℂ) :=
+{ carrier := is_cusp_form k Γ,
+  zero_mem' := { hol := (λ _, mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ)),
+    transf := (weakly_modular_form k Γ).zero_mem',
+    infinity :=  λ  A, begin
+    rw slash_action.mul_zero,
+    apply (zero_at_im_infty_submodule ℂ).zero_mem
+    end},
+  add_mem' := λ a b ha hb,
+    { hol := mdifferentiable.add _ ha.hol hb.hol,
+      transf := (weakly_modular_form k Γ).add_mem' ha.transf hb.transf,
+      infinity := λ A, begin
+      rw slash_action.add_action,
+      apply ((zero_at_im_infty_submodule ℂ).add_mem' (ha.infinity A) (hb.infinity A))
+    end},
+  smul_mem' :=λ c f hf,
+    { hol :=  mdifferentiable.const_smul _ _ hf.hol,
+      transf := (weakly_modular_form k Γ).smul_mem' _ hf.transf,
+      infinity := λ A, begin
+      rw slash_action.smul_action,
+      apply (zero_at_im_infty_submodule ℂ).smul_mem' c (hf.infinity A),
+    end}}
+
+localized "notation `S`:= space_of_cusp_forms_of_weight_and_level" in modular_forms
+
+/-- This is the space of modular forms of weight `k` and level `Γ`-/
+def space_of_mod_forms_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2, ℤ)) : submodule ℂ (ℍ → ℂ) :=
+{ carrier := { f : ℍ → ℂ | is_modular_form k Γ f },
+  zero_mem':= is_modular_form_of_is_cusp_form (S k Γ).zero_mem',
+  add_mem' := λ  a b ha hb,
+    { hol := mdifferentiable.add _ ha.hol hb.hol,
+      transf :=  (weakly_modular_form k Γ).add_mem' ha.transf hb.transf,
+      infinity := λ  A, begin
+      rw slash_action.add_action,
+      exact (bounded_at_im_infty_subalgebra ℂ).add_mem' (ha.infinity A) (hb.infinity A)
+      end},
+  smul_mem' := λ c f hf,
+    { hol := mdifferentiable.const_smul _ _ hf.hol,
+      transf := (weakly_modular_form k Γ).smul_mem' _ hf.transf,
+      infinity := λ A, begin
+      rw slash_action.smul_action,
+      apply (bounded_at_im_infty_subalgebra ℂ).smul_mem (hf.infinity A)
+    end } }
+
+localized "notation `M`:= space_of_mod_forms_of_weight_and_level " in modular_forms
+
+/--The product of two modular forms is a modular form whose weight is the sum of the weights-/
+lemma mul_modform (k_1 k_2 : ℤ) (Γ : subgroup SL(2, ℤ)) (f g : ℍ → ℂ)
+  (hf : f ∈ M k_1 Γ) (hg : g ∈ M k_2 Γ) : f * g ∈ (M (k_1 + k_2) Γ) :=
+{ hol := mdifferentiable.mul _ hf.hol hg.hol,
+  transf := mul_modular _ _ _ _ _ hf.transf hg.transf,
+  infinity := λ A, begin
+  rw [slash_mul_SL2 k_1 k_2 A f g],
```

Comment:
```suggestion
  rw slash_mul_SL2,
```
You often fill in information in rewrites that is not needed.  This is just a matter of taste, so it's your choice, but personally I find it more readable with just the lemma name.